### ISSUE TITLE
Child process is exited prematurely when running a view command

### ISSUE THREAD

**Issue by ncuillery** [npm/npm#16780]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [x] Other (_see below for feature requests_):

#### What's going wrong?

When using NPM 5: if I run the `npm view` command in a child process (either `exec` or `spawn`) the child process exits with code 0 right after the reception of the first chunk of data.

#### How can the CLI team reproduce the problem?

Considering this simple Node program:
```javascript
const exec = require('child_process').exec;

const child = exec('npm view npm@latest');

child.stdout.on('data', data => {
  console.log('DATA', data.length);
});

child.stderr.on('data', data => {
  console.log('ERROR', data.length);
});

child.on('exit', code => {
  console.log('EXIT', code);
});
```

Output when NPM 4.6.1 is installed:
```
$ node debug-exec.js      
DATA 8192
DATA 8192
DATA 8189
DATA 8191
DATA 3817
EXIT 0
```

Output when NPM 5.0.0 is installed:
```
$ node debug-exec.js      
DATA 8192
EXIT 0
```

There is no problem with these commands (the whole output is received whatever the NPM version):
 - `ls -l node_modules`
 - `curl -s https://github.com/ncuillery`

### supporting information:

 - `npm -v` prints: either `4.6.1` or `5.0.0`
 - `node -v` prints: `v7.1.0`
 - `npm config get registry` prints: `https://registry.npmjs.org/`
 - Windows, OS X/macOS, or Linux?: macOS Sierra 10.12.4
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by ncuillery** [comment]
Additional notes:

#### Same bug using the exec's callback argument:

```javascript
const exec = require('child_process').exec;

exec('npm info npm@latest', (err, stdout, stderr) => {
  console.log('DATA', (stdout || []).length);
  console.log('ERROR', (stderr || []).length);
});
```


Output with NPM 4.6.1
```
$ node debug-exec2.js
DATA 36581
ERROR 0
```

Output with NPM 5.0.0
```
$ node debug-exec2.js
DATA 8192
ERROR 0
```

#### Same bug with a `yarn` command regardless of the installed NPM version

```
$ node debug-exec-yarn.js      
DATA 18
DATA 8192
EXIT 0
```
The first chunk of 18b corresponds to the repetition of the command by yarn.

**Comment by zkat** [comment]
This is so weird. We didn't really change anything about `npm view` in npm5 😂 

Thanks for the report, we'll take a look soon 👍

**Comment by evanlucas** [comment]
This seems to be due to 30d79062869c49c0d6cf6adf57cfbe79c2c050b9. 

This patch seems to work from a quick test:

```
diff --git a/lib/utils/error-handler.js b/lib/utils/error-handler.js
index 5374d1f..9e07961 100644
--- a/lib/utils/error-handler.js
+++ b/lib/utils/error-handler.js
@@ -135,7 +135,7 @@ function exit (code, noLog) {
     // for whatever reason gets thrown away, instead of leaving the CLI open
     //
     // Commands that expect long-running actions should just delay `cb()`
-    process.exit(code)
+    process.exitCode = code
   }
 }
 ```

@zkat was the explicit `process.exit()` added to account for outstanding requests/timers or something like that?

**Comment by zkat** [comment]
@evanlucas indeed. There's been the occasional thing, sometimes on Windows, where some mystery handle stays open and keeps the npm process alive. We thought outright exiting here would be alright, but apparently not always. I think I've run into other situations where this is being held up.

I'd much rather hunt down the parts of the code that are relying on dangling processes and explicitly callback on them exiting. :\

The more immediate, consistent issue is that without this change, anyone using proxies is going to have their npm process hang on exit, due to https://github.com/zkat/make-fetch-happen/issues/29

**Comment by zkat** [comment]
Found it. PR Incoming.

**Comment by zkat** [comment]
The fix will be included in `5.0.1`. node@8 already hotfixed this in their version of `5.0.0` (see https://github.com/MylesBorins/node/commit/c5c68d7857631a539e2ee65eb8f95215487dea42)
